control/pure_pursuit/src/pure_pursuit/pure_pursuit_lateral_controller.cpp

```diff
@@ -99,42 +120,301 @@ bool PurePursuitLateralController::isDataReady()
+    param_.max_lookahead_distance);
+
+  auto pubDebugValues = [&]() {
+    autoware_auto_system_msgs::msg::Float32MultiArrayDiagnostic debug_msg{};
+    debug_msg.diag_array.data.resize(TYPE::SIZE);
+    debug_msg.diag_array.data.at(TYPE::VEL_LD) = static_cast<float>(vel_ld);
+    debug_msg.diag_array.data.at(TYPE::CURVATURE_LD) = static_cast<float>(curvature_ld);
+    debug_msg.diag_array.data.at(TYPE::LATERAL_ERROR_LD) = static_cast<float>(lateral_error_ld);
+    debug_msg.diag_array.data.at(TYPE::TOTAL_LD) = static_cast<float>(total_ld);
+    debug_msg.diag_array.data.at(TYPE::VELOCITY) = static_cast<float>(velocity);
+    debug_msg.diag_array.data.at(TYPE::CURVATURE) = static_cast<float>(curvature);
+    debug_msg.diag_array.data.at(TYPE::LATERAL_ERROR) = static_cast<float>(lateral_error);
+    debug_msg.diag_header.data_stamp = node_->now();
+    pub_debug_values_->publish(debug_msg);
+  };
+
+  if (is_control_cmd) {
+    pubDebugValues();
+  }
+
+  return total_ld;
+}
+
 void PurePursuitLateralController::setInputData(InputData const & input_data)
 {
   trajectory_ = input_data.current_trajectory_ptr;
   current_odometry_ = input_data.current_odometry_ptr;
   current_steering_ = input_data.current_steering_ptr;
 }
 
+TrajectoryPoint PurePursuitLateralController::calcNextPose(
+  const double ds, TrajectoryPoint & point, AckermannLateralCommand cmd) const
+{
+  geometry_msgs::msg::Transform transform;
+  transform.translation = tier4_autoware_utils::createTranslation(ds, 0.0, 0.0);
+  transform.rotation =
+    planning_utils::getQuaternionFromYaw(((tan(cmd.steering_tire_angle) * ds) / param_.wheel_base));
+  TrajectoryPoint output_p;
+
+  tf2::Transform tf_pose;
+  tf2::Transform tf_offset;
+  tf2::fromMsg(transform, tf_offset);
+  tf2::fromMsg(point.pose, tf_pose);
+  tf2::toMsg(tf_pose * tf_offset, output_p.pose);
+  return output_p;
+}
+
+void PurePursuitLateralController::setResampledTrajectory()
+{
+  // Interpolate with constant interval distance for lateral acceleration calculation.
+  std::vector<double> out_arclength;
+  const auto input_tp_array = motion_utils::convertToTrajectoryPointArray(*trajectory_);
+  const auto traj_length = motion_utils::calcArcLength(input_tp_array);
+  for (double s = 0; s < traj_length; s += param_.resampling_ds) {
+    out_arclength.push_back(s);
+  }
+  trajectory_resampled_ =
+    std::make_shared<autoware_auto_planning_msgs::msg::Trajectory>(motion_utils::resampleTrajectory(
+      motion_utils::convertToTrajectory(input_tp_array), out_arclength));
+  trajectory_resampled_->points.back() = trajectory_->points.back();
+  trajectory_resampled_->header = trajectory_->header;
+  output_tp_array_ = boost::optional<std::vector<TrajectoryPoint>>(
+    motion_utils::convertToTrajectoryPointArray(*trajectory_resampled_));
+  if (
+    param_.enable_path_smoothing &&
+    (trajectory_resampled_->points.size() > 2 * param_.path_filter_moving_ave_num)) {
+    averageFilterTrajectory(*trajectory_resampled_);
```

Comment:
No one can expect `averageFilterTrajectory` is applied in the `setResampledTrajectory` function from its name. `setResampledTrajectory` should work only for the resampling. The `averageFilterTrajectory` should be applied somewhere else. 

I think it can be moved to the `run` function.